synapse/handlers/message.py

```diff
@@ -989,10 +1019,12 @@ async def create_new_client_event(
             context = EventContext.for_outlier()
         elif (
             event.type == EventTypes.MSC2716_INSERTION
-            and full_state_ids_at_event
+            and state_event_ids
             and builder.internal_metadata.is_historical()
         ):
-            old_state = await self.store.get_events_as_list(full_state_ids_at_event)
+            # Add explicit state to the insertion event so the rest of the batch
+            # can inherit the same state and state_group
+            old_state = await self.store.get_events_as_list(state_event_ids)
```

Comment:
Added some nice context for why we're doing this that was only a PR review comment previously, https://github.com/matrix-org/synapse/pull/10975#discussion_r721867239